common/scala/src/main/scala/org/apache/openwhisk/http/PoolingRestClient.scala

```diff
@@ -110,9 +112,17 @@ class PoolingRestClient(
       if (response.status.isSuccess) {
         Unmarshal(response.entity.withoutSizeLimit).to[T].map(Right.apply)
       } else {
-        // This is important, as it drains the entity stream.
-        // Otherwise the connection stays open and the pool dries up.
-        response.discardEntityBytes().future.map(_ => Left(response.status))
+        Unmarshal(response.entity).to[JsObject].flatMap { responseEntity =>
+          val error = responseEntity.fields.get("error").getOrElse("").toString
+          val reason = responseEntity.fields.get("reason").getOrElse("").toString
+          val statusCode = response.status
+          val statusCodeReason = statusCode.reason()
+          val customReason = s"$statusCodeReason (error: $error, reason: $reason)"
+          val customStatusCode = StatusCodes.custom(intValue = statusCode.intValue(), reason = customReason, defaultMessage = statusCode.defaultMessage())
```

Comment:
It feels... weird to overload StatusCode with more meaning than is already in the HTTP spec. Should we adjust the future to return `Either[ExpressiveErrorType, T]` rather than just statuscode?

Reply:
> It feels... weird to overload StatusCode with more meaning than is already in the HTTP spec. Should we adjust the future to return `Either[ExpressiveErrorType, T]` rather than just statuscode?

But wouldn't that be an interface change implying changes in many classes (including tests) doing REST calls using `requestJson` in `PoolingRestClient` as implementation?

Reply:
It would, yes.

Reply:
Would you both agree that having this info available in the log is valuable? 

Regarding defining an own error type, as this is a breaking change not only in openwhisk, would you prefer it anyway (the implementation only extends the reason field, but the new StatusCode is derived from the orginal one)? Another possibility would be to simply log the valuable info from the response entity in method requestJson in case of an error.

Should be go forward with this PR and which way do you suggest?

Reply:
Given all the other comments, I'd say it's okay to move forward this way.

However, I would not parse the response as JSON but as a String. That way you catch many more use-cases than just the very narrow one outlined here. WDYT?

Reply:
+1 there’s no guarantee response is json.

Reply:
Yes makes sense, I changed the implementation to parse String instead of JSON. For the discussed error case the error now looks as follow:

`Unexpected http response code: 400 Bad Request (details:{"error":"query_parse_error","reason":"Limit is too large, must not exceed 268435456"} ) [marker:database_queryView_error:402:33]`

Reply:
LGTM. you can remove `()` on `.intValue()` and `.defaultMessage()`.